ide/gnatstudio/recordflux.py

```diff
@@ -0,0 +1,268 @@
+"""
+RecordFlux support for GNAT Studio
+"""
+
+import os.path
+import re
+
+import GPS
+import highlighter.common as hl
+from gs_utils import hook
+
+XML = r"""<?xml version="1.0"?>
+<GNAT_Studio>
+   <Language>
+      <Name>RecordFlux</Name>
+      <Spec_Suffix>.rflx</Spec_Suffix>
+
+      <!-- Dummy body suffix as file templates don't work when its missing -->
+      <Body_Suffix>.workaround_rflx</Body_Suffix>
+
+      <Categories>
+         <Category>
+            <Name>package</Name>
+            <Pattern>^[ \\t]*package[ \\t]+((\\w|\\.)+)</Pattern>
+            <Index>1</Index>
+         </Category>
+         <Category>
+            <Name>type</Name>
+               <Pattern>^[ \\t]*type[ \\t]+(\\w+)</Pattern>
+               <Index>1</Index>
+         </Category>
+      </Categories>
+
+      <Context>
+         <New_Line_Comment_Start>--</New_Line_Comment_Start>
+         <String_Delimiter>&quot;</String_Delimiter>
+         <Can_Indent>True</Can_Indent>
+         <Syntax_Highlighting>True</Syntax_Highlighting>
+         <Case_Sensitive>False</Case_Sensitive>
+      </Context>
+   </Language>
+
+   <!-- Filter -->
+   <filter name="RecordFlux" language="RecordFlux"/>
+
+   <!-- Actions -->
+   <action name="check">
+      <filter_and>
+         <filter id="RecordFlux"/>
+         <filter id="Source editor"/>
+      </filter_and>
+      <external>rflx check %F</external>
+      <on-failure>
+         <shell lang="python" show-command="false">recordflux.parse_output(&quot;&quot;&quot;%1&quot;&quot;&quot;)</shell>
+      </on-failure>
+   </action>
+
+   <action name="check_all">
+      <filter id="File"/>
+      <shell lang="python" show-command="false">recordflux.get_source_files()</shell>
+      <external>rflx check %1"</external>
+      <on-failure>
+         <shell lang="python" show-command="false">recordflux.parse_output(&quot;&quot;&quot;%1&quot;&quot;&quot;)</shell>
+      </on-failure>
+   </action>
+
+   <action name="generate">
+      <filter_and>
+         <filter id="RecordFlux"/>
+         <filter id="Source editor"/>
+      </filter_and>
+      <external>rflx generate -d generated %F</external>
+      <on-failure>
+         <shell lang="python" show-command="false">recordflux.parse_output(&quot;&quot;&quot;%1&quot;&quot;&quot;)</shell>
+      </on-failure>
+   </action>
```

Comment:
After starting GNAT Studio with `gnatstudio -Ptests/ide/recordflux.gpr`, the `Generate` action does not work:

```Console
rflx generate -d generated /[...]/RecordFlux/specs/tlv.rflx
cli: error: directory not found: "generated"
```

Where does it look for the `generated` directory? Is it possible to define the output directory in the project file?

Reply:
In fact it's hardcoded and the location of the GPR file is used as a base directory (when you do `mkdir tests/ide/generated` files are generated there).

We should decide how to handle that. Maybe we should generate into a specific element of `Source_Dirs`, as the generated code will probably be used by GNAT/SPARK from there later. Question is how we decide which directory this should be (special name? additional project file parameter that defines it?). I don't know whether custom project file attributes can be used, need to look into that.

Reply:
I solved the issue in d899eb6d14e8e68e9762067b2c73cd48b81a4a95. The RecordFlux output directory can be configured in the project file now:

```Ada
project RecordFlux is                                                                                                                                                         
                                                                                                                        
   for Create_Missing_Dirs use "True";                                                                                  
   for Source_Dirs use (".", "../../specs");                                                                            
   for Languages use ("Ada", "RecordFlux");                                                                             
                                                                                                                        
   package Recordflux is                                                                                                
      for Output_Dir use "../../generated";                                                                             
   end Recordflux;                                                                                                      
                                                                                                                        
end RecordFlux;
```

If `Create_Missing_Dirs` is set, the directory is created prior to running `rflx`. The `Output_Dir` property can also be configured in the RecordFlux pane of the project properties in GNAT Studio:

![RecordFlux_GNAT_Studio_Properties](https://user-images.githubusercontent.com/6357197/85926200-09fadb00-b89e-11ea-8c02-a878e95b70ab.png)